src/lsm/segmented_array.zig

```diff
@@ -120,16 +153,55 @@ fn SegmentedArrayType(
             allocator.free(array.indexes);
         }
 
+        pub fn verify(array: Self) void {
+            for (array.nodes) |node, node_index| {
+                if (node_index < array.node_count) {
+                    // The first node_count pointers are non-null.
+                    assert(node != null);
+                } else {
+                    // The rest are non-null.
+                    assert(node == null);
+                }
+            }
+            for (array.nodes[0..array.node_count]) |_, node_index| {
+                const c = array.count(@intCast(u32, node_index));
+                // Every node is at most full.
+                assert(c <= node_capacity);
+                // Every node is at least half-full, except the last.
+                if (node_index < array.node_count - 1) {
+                    assert(c >= @divTrunc(node_capacity, 2));
+                }
+            }
+            if (Key) |K| {
+                // If Key is not null then the elements must be sorted by key_from_value (but not necessarily unique).
+                var last_key: ?K = null;
+                for (array.nodes[0..array.node_count]) |_, node_index| {
+                    for (array.node_elements(@intCast(u32, node_index))) |*value| {
+                        const key = key_from_value(value);
+                        if (last_key != null) {
+                            assert(compare_keys(last_key.?, key) != .gt);
+                        }
```

Comment:
```suggestion
                        if (last_key) |last| {
                            assert(compare_keys(last, key) != .gt);
                        }
```

Reply:
We could also inline the assert?

```zig
                        if (last_key) |last| assert(compare_keys(last, key) != .gt);
```

Reply:
Do you have thoughts on naming conventions for nullable stuff. I've often done `if (last_key_maybe) |last_key| ...`.

Reply:
I don't think we have a convention for it. I like the idea of a common suffix, but I think "maybe" as a suffix is a little misleading, since e.g. `key_prior_maybe` is _always_ the prior key, which may or may not exist. :thinking:

Reply:
`key_prior_if_exists`? `key_prior_or_null`?

Reply:
It's fine also not to include the optionality in the name, and just use `key_prior`, since it's shown both by the type system, and when doing `if (key_prior)` or unwrapping via `.?` so the suffix is redundant.

Reply:
It's just the name shadowing - `if (key_prior) |key_prior| ...` works in rust but not in zig. I used `key_prior_o` in my own code but somehow think that isn't going to fly in TB :)

Reply:
Ah great, I was wondering whether you were thinking about shadowing.

We could do `key_prior_exists` then so that it reads `if (key_prior_exists) |key_prior|` or else just capture as `key` or `prior` depending on scope, as you prefer.